src/lib/reasoners/intervalCalculus.ml

```diff
@@ -2082,20 +2085,95 @@ let model_from_unbounded_domains =
     let l2 = model_from_simplex int_sim true  env uf in
     List.fold_left mk_cs (List.fold_left mk_cs [] l1) l2
 
+let mk_const_term c ty =
+  let c = Q.to_string c in
+  match ty with
+  | Ty.Tint -> E.int c
+  | Ty.Treal -> E.real c
+  | _ -> assert false
 
-let case_split env uf ~for_model =
-  let res = default_case_split env uf ~for_model in
-  match res with
-  | [] ->
-    if not for_model then []
+let optimizing_split env uf opt =
+  (* soundness: if there are expressions to optmize, this should be
+     done without waiting for ~for_model flag to be true *)
+  let {Th_util.order; r = r; is_max = to_max; e; value } = opt in
+  assert (match value with
+      | Unknown -> true
+      | _ -> false
+    );
+  let repr, _ = Uf.find uf e in
+  let ty = E.type_info e in
+  let r1 = r in (* instead of repr, which may be a constant *)
+  let p = poly_of repr in
+  match P.is_const p with
+  | Some optim ->
+    Format.eprintf
+      "%a has a %s: %a@."
+      E.print e
+      (if to_max then "maximum" else "minimum")
+      Q.print optim;
+    let r2 = alien_of (P.create [] optim  ty) in
+    let t2 = mk_const_term optim ty in
+    Debug.case_split r1 r2;
+    let o = Some {Th_util.opt_ord = order; opt_val = Th_util.Value t2} in
+    let s = LR.mkv_eq r1 r2, true, Th_util.CS (o, Th_util.Th_arith, Q.one) in
+    Sig_rel.Optimized_split { opt with value = Value s; }
+
+  | None ->
+    let sim = if ty == Ty.Tint then env.int_sim else env.rat_sim in
+    let p = if to_max then p else P.mult_const Q.m_one p in
+    let l, c = P.to_list p in
+    let l = List.rev_map (fun (x, y) -> y, x) (List.rev l) in
+    let sim, mx_res = Sim.Solve.maximize sim (Sim.Core.P.from_list l) in
+    match Sim.Result.get mx_res sim with
+    | Sim.Core.Unknown -> assert false
+    | Sim.Core.Sat _   -> assert false (* because we maximized *)
+    | Sim.Core.Unsat _ -> assert false (* we know sim is SAT *)
+    | Sim.Core.Unbounded _ ->
+      Format.eprintf
+        "%a is unbounded. Let other methods assign a value for it@."
+        E.print e;
+      let value = if to_max then Th_util.Pinfinity else Th_util.Minfinity in
+      Sig_rel.Optimized_split { opt with value }
+
+    | Sim.Core.Max(mx,_sol) ->
+      let {Sim.Core.max_v; _} = Lazy.force mx in
```

Comment:
I am surprised we ignore `is_le` here. In fact, Alt-Ergo loops when given an objective that has a strict bound (e.g. maximize `x` under the constraint `x < 0.5`).

Reply:
I fixed by giving up optimization when I got such optimization constraints.